library/std/src/os/windows/io/handle.rs

```diff
@@ -200,23 +200,57 @@ impl OwnedHandle {
 }
 
 impl TryFrom<HandleOrInvalid> for OwnedHandle {
-    type Error = ();
+    type Error = InvalidHandleError;
 
     #[inline]
-    fn try_from(handle_or_invalid: HandleOrInvalid) -> Result<Self, ()> {
+    fn try_from(handle_or_invalid: HandleOrInvalid) -> Result<Self, InvalidHandleError> {
         let owned_handle = handle_or_invalid.0;
         if owned_handle.handle == c::INVALID_HANDLE_VALUE {
             // Don't call `CloseHandle`; it'd be harmless, except that it could
             // overwrite the `GetLastError` error.
             forget(owned_handle);
 
-            Err(())
+            Err(InvalidHandleError(()))
         } else {
             Ok(owned_handle)
         }
     }
 }
 
+/// This is the error type used by [`HandleOrNull`] when attempting to convert
+/// into a handle, to indicate that the value is null.
+#[unstable(feature = "io_safety", issue = "87074")]
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NullHandleError(());
```

Comment:
Are you intentionally adding a private `()` field here to ensure this isn't constructable?

Reply:
Yes; `std` does this in several places, such as `CharTryFromError`, `AddrParseError`, and `FromBytesUntilNulError`. It's not strictly necessary here, but it seems reasonable to do, to restrict the API surface area as much as possible.

Reply:
Sounds reasonable. r=me with a brief mention in a (non-doc) comment here, something like `// The empty field prevents constructing this, and allows extending it in the future`.

Reply:
I've now added these comments.